### DISCUSSION TITLE
max3421e <> Flipper Zero (STM32WB55RG)

### DISCUSSION THREAD

**Discussion by zacharyweiss** [hathach/tinyusb#2623] [Q&A]
Howdy!

I'm new to embedded development, and currently drinking from the firehose of topics, so apologies if this is a misguided line of questions.

I'm interested in building a USB host module for the Flipper Zero (based on the STM32WB55RG), and the max3421e seems a fitting option for this, which I see this project has [implemented code for](https://github.com/hathach/tinyusb/blob/master/src/portable/analog/max3421/hcd_max3421.c). As far as I understand from a skim of the docs / discussions / code, it seems you implement the SPI functions on a per-hardware basis in `hw/bsp/{NAME}/family.c` (eg, for the [RP2040](https://github.com/hathach/tinyusb/blob/104a5daef49d91b54f7af499090efd478c936971/hw/bsp/rp2040/family.c)). 

In Flipper-development-land, short of adding this to the firmware, apps only have access to a subset of wrapped `hal` functions; hence, the analogous SPI API funcs I would need to write for use in a Flipper App would be a bit different than using the low-level HAL of the STM32WB55RG itself. Given this, what would be the most appropriate way for me to implement the SPI API funcs for your max3421e code? Any way to write it that would conform to the existing approach of tinyusb for different HW, or best to cannibalize just the `src/portable/analog/max3421/hcd_max3421.c` file and implement those SPI API funcs locally in my project using the Flipper's API? (and more broadly, am I asking the right questions here, or misunderstanding the proper approach of using tinyusb?)

**Comment by zacharyweiss** [comment]
e.g., would it be appropriate to create a "sub-board" file in `hw/bsp/stm32wb/` that specifies use of the wrapped / Flipper-Zero-Firmware-specific functions, a la `BOARD=STM32WB_FLIPPERZERO`?

**Reply by hathach** [reply]
I am not familliar with flipeer0-dev-land, but for max3421 host to work you need to implement
- spi API to make transfer
- gpio interrupt to call hcd_int_handler() which process interrupt event

sure you can add new flipper zero board, implement above API in family.c, if it is too specific for flipper0, you can use #ifdef BOARD_STM32WB_FLIPPERZERO (which is added by example [makefile](https://github.com/hathach/tinyusb/blob/master/examples/build_system/make/make.mk#L110))
